code/layout-analyzer.js

```diff
@@ -31,32 +31,101 @@ window.addEventListener('DOMContentLoaded', () => {
 
   const charToKeys = char => keyChars[char] ?? keyChars[substituteChars[char]];
 
+  const is1DFH = keyCode =>
+    keyCode.startsWith('Key') ||
+      ['Space', 'Comma', 'Period', 'Slash', 'Semicolon'].includes(keyCode);
+
   // create an efficient hash table to parse a text
   const supportedChars = (keymap, deadkeys) => {
     const charTable = {};
     const deadTable = {};
 
-    // main chars, directly accessible (or with Shift / AltGr)
-    Object.entries(keymap).forEach(([key, values]) => {
-      values.forEach(char => {
-        if (!(char in charTable)) {
-          if (char.length === 1) {
-            charTable[char] = [key];
-          } else if (!(char in deadTable)) {
-            deadTable[char] = key;
-          }
-        }
-      });
-    });
+    // In case there are multiple ways of typing a singel char, this checks
+    // which sequence is easier to type (examples are in Ergo‑L)
+    const requiresLessEffort = (originalKeySequence, newKeySequence) => {
+      const uses1DK = '**' in keyboard.layout.deadKeys
+          ? keySequence => keySequence.some(key => key == charTable['**'][0])
+          : (_) => false;
+
+      const arrayCount = (array, predicate) => {
+        let rv = 0;
+        array.forEach(elem => { if (predicate(elem)) rv++ });
+        return rv;
+      };
+
+      const cmp = (val1, val2) => {
+        if (val1 > val2) return "more";
+        if (val1 < val2) return "less";
+        return "same";
+      };
+
+      if (originalKeySequence.length > 1 && newKeySequence.length > 1) {
+        const cmp1DK = cmp(
+          uses1DK(originalKeySequence),
+          uses1DK(newKeySequence)
+        );
+        if (cmp1DK == 'more') return false;
+        if (cmp1DK == 'less') return true;
```

Comment:
Nitpick, but I’m surprised your linter hasn’t yelled at you for using `==` instead of `===`

Reply:
Yet another nitpick, please swap these two lines (consistency with the rest of the file)